### ISSUE TITLE
In OSSL thread pool, side effects when setting max thread number per libctx

### ISSUE THREAD

**Issue by docularxu** [openssl/openssl#20059]
This is a bug report following our discussions in https://github.com/openssl/openssl/pull/12256#discussion_r1069159301  : 
The OSSL_set_max_threads() call to set a limit per libctx.

I see one issue in the current thread pool design. When two (or more) users in one libctx needs to use thread pool, then they must coordinate in how many max threads should be set. However this is not the end of story. Even they reach an agreement on the maximum number of threads to set in this libctx, they met the problem of stealing threads from each other when calling ossl_crypto_thread_start().

Eg. user A wants 5 threads, and user B wants 5 threads, and they are both in one libctx. So, first we OSS_set_max_threads() to 10. Then, user A spawns tasks with ossl_crypto_thread_start(). If the limit is 5, user A's spawn should wait on _ossl_get_avail_threads() until other tasks being free'ed.  Ref: https://github.com/openssl/openssl/blob/ae61bd34e676036f22e44e8184d1ba649a990619/crypto/thread/internal.c#L55

However, here in the multi-user case, the max limit is set to 10. So, user A can spawn more than 5 threads. And when user B starts to spawn her threads, it's quite possible, there is less left than what user B expected.

@mattcaswell reported another issue in the thread pool deign, which is the end user can lose control over the "total" size of the thread pool. Ref: https://github.com/openssl/openssl/pull/12256#discussion_r1069159301

**Comment by ckalina** [comment]
For reference purposes: this would likely require changes to the default thread pool design as specified in https://github.com/openssl/openssl/pull/18864, merged in https://github.com/openssl/openssl/pull/19455.
